Beagle Core Stubs

jGleitz opened the pull request:
Created stubs for Beagle Core.

 * Most is as discussed
 * Added two Exceptions I thought made sense 
 * Added a mechanism to store custom data to the blackboard

fixes #298

jGleitz commented:
fixes #298

RomanLangrehr commented on .settings/org.eclipse.jdt.ui.prefs:
```diff
@@ -65,4 +65,4 @@ org.eclipse.jdt.ui.importorder=;de;org;com;java;javax;
 org.eclipse.jdt.ui.ondemandthreshold=99
 org.eclipse.jdt.ui.staticondemandthreshold=99
-org.eclipse.jdt.ui.text.custom_code_templates=<?xml version\="1.0" encoding\="UTF-8" standalone\="no"?><templates><template autoinsert\="false" context\="gettercomment_context" deleted\="false" description\="Comment for getter method" enabled\="true" id\="org.eclipse.jdt.ui.text.codetemplates.gettercomment" name\="gettercomment">/**\n * TODO describe ${bare_field_name}\n * \n * @return the ${bare ...
+org.eclipse.jdt.ui.text.custom_code_templates=<?xml version\="1.0" encoding\="UTF-8" standalone\="no"?><templates><template autoinsert\="false" context\="gettercomment_context" deleted\="false" description\="Comment for getter method" enabled\="true" id\="org.eclipse.jdt.ui.text.codetemplates.gettercomment" name\="gettercomment">/**\n * TODO describe \u201C${bare_field_name}\u201D.\n * \n * @retu ...
```
What happened here?

(Just to make sure, this will not be changed with every change).

jGleitz replied:
I adapted the code templates to be more meaningful and produce less checkstyle errors (51f388a). The changes are intended.

RomanLangrehr commented on src/main/java/de/uka/sdq/beagle/core/Blackboard.java:
```diff
@@ -0,0 +1,106 @@
+	 * always pass its own {@linkplain Class} instance as {@code writer}. {@code writer}
+	 * has to implement {@link BlackboardStorer} and thereby declare the type of values
+	 * they write. Calling this method will override any data potentially stored
+	 * previously for the given {@code writer}.
```
I'm not sure, whether this is the a good idea. I rather think most `ResultAnalyser` will rather like to add informations than replacing them.

jGleitz replied:
Look at the example below. It a `ResultAnalyser` wants to add data, he can simply store a collection.

RomanLangrehr commented on src/main/java/de/uka/sdq/beagle/core/BlackboardStorer.java:
```diff
@@ -0,0 +1,13 @@
+ *            The type of data the BlackboardStorer wants to write to the blackboard.
+ */
+public interface BlackboardStorer<WRITTEN_TYPE extends Serializable> {
+}
```
I didn't get the sense of this interface yet. Each user, that want's to store something special on the blackbord has to create a class (for each object he want's to store!), which implements this empty interface ?!

jGleitz replied:
See the example below. It’s a lot easier than that ;)

RomanLangrehr commented on src/main/java/de/uka/sdq/beagle/core/expressions/DivisionExpression.java:
```diff
@@ -0,0 +1,18 @@
+ * TODO Document this type.
+ *
+ */
+public class DivisionExpression {
```
`implements Evaluable Expression` is missing, insn't it?

jGleitz replied:
yep. Fixed it, thank you.

RomanLangrehr commented on src/main/java/de/uka/sdq/beagle/core/expressions/EvaluableExpressionVisitor.java:
```diff
@@ -0,0 +1,22 @@
+
+	void visit(MultiplicationExpression expression);
+
+	void visit(EvaluableVariable variable);
```
Is it on purpose, that some methods are missing here?

jGleitz replied:
No, I simply forgot about them. Thanks for reminding me! I fixed it.

jGleitz commented:
As I see that my construct for storing data on the blackboard is not clear at first sight, I’ll make an example. If we decide to stick to the concept, I’ll add the example to the Javadoc.

Please note that while the interfaces may **look** a little complicated, they are *very easy* to **use**.

Let’s say we have 2 analysers, `MyAnalyser` and `YourAnalyser`. Both want to store data on the Blackboard to keep track whether there is something new to look at.

**`YourAnalyser`** simply wants to keep track of the `SEFFLoop`s he has already seen. He wants to use an `HashSet` to do so:

```java
public class YourAnalyser implements ResultAnalyser, BlackboardStorer<HashSet<SEFFLoop>> {
    
    public boolean canContribute(ReadOnlyBlackboardView blackboard) {
        Set<SEFFLoop> alreadySeen = blackboard.readFor(YourAnalyser.class);
        Set<SEFFLoop> allLoops = blackboard.getAllSEFFLoops();
        return allLoops.size() > 0 && (alreadySeen == null || !alreadySeen.containsAll(allLoops));
    }

    public boolean contribute(AnalyserBlackboardView blackboard) {
        Set<SEFFLoop> alreadySeen = blackboard.readFor(YourAnalyser.class);
        if (alreadySeen == null) {
            alreadySeen = new HashSet<SEFFLoop>();
            blackboard.writeFor(YourAnalyser.class, alreadySeen);
        }
        Set<SEFFLoop> allLoops = blackboard.getAllSEFFLoops();
        Set<SEFFLoop> todo = allLoops.removeAll(alreadySeen);
        
        for (SEFFLoop loop : todo) {
             // do the logic here
             alreadySeen.add(loop);
        }
    }
}
```
Simple, isn’t it?


**`MyAnalyser`** however, wants to use its own data structure (which will typically be the case, I think):

```java
public MyAnalyserDataStructure implements Serializable {
   // a whole bunch of getters and setters
}
```


```java
public class MyAnalyser implements ResultAnalyser, BlackboardStorer<MyAnalyserDataStructure> {
    // well, everything here is like above:
    public boolean canContribute(ReadOnlyBlackboardView blackboard) {
        MyAnalyserDataStructure stored = blackboard.readFor(MyAnalyser.class);
        // …
    }

    public boolean contribute(AnalyserBlackboardView blackboard) {
        MyAnalyserDataStructure stored = blackboard.readFor(MyAnalyser.class);
        // …
        MyAnalyserDataStructure newData = new MyAnalyserDataStructure();
        // …
        blackboard.writeFor(MyAnalyser.class, newData);
        // …
    }
```

I think this makes storing and reading really simple.

RomanLangrehr commented:
But there is a problem, when an analyser want's to store multiple objects. He has to create a lot of empty classes and can't adapt it during runtime.

I don't think it's a good idea to use classes as keys. I think strings are better: When we specify that the strings must have the form `my.package.whateverIWantJustWithoutDots` they are even unique. Additionally it is easy to share data between two analysers.

jGleitz commented:
First of all, I feel like there is a misunderstanding here: When reading and writing data, an object does not pass the class of the object he wants to read/write, but its *own* class! It declared the type of the data he wants to read and write by implementing `BlackboardStorer`. Usually, it will write its own, serialisable class to hold all its data and declare that as the type it wants to read and write.

> But there is a problem, when an analyser want's to store multiple objects.

Well it would just build a custom data class. That’s not a big deal, is it?

> He has to create a lot of empty classes

No, just one data class (That contains some fields, getters, and setters and thus is all but “empty”)

> and can't adapt it during runtime.

All it has to know is the types of the objects it’ll want to use so it can build the data class. And as Java is a type safe language, it’s reasonable to expect that.

>  I think strings are better

One major downside: You completely lose **type safety**. You’ll have to use a lot of casts. Actually, type safety is all this interface is about.

> Additionally it is easy to share data between two analysers.

First of all: Sharing data between analysers *completly contradicts our architecture*. However, you can still easily get that with my system. And you’ll have **type safety**!

RomanLangrehr commented:
Lgtm.

RomanLangrehr commented on src/main/java/de/uka/sdq/beagle/analysis/ResultAnalyser.java:
```diff
@@ -0,0 +1,52 @@
+	 * that its result do not depend on any prior calls to any method (especially not
+	 * {@link #canContribute(ReadOnlyBlackboardView)}) but only on the passed
+	 * {@code blackboard}. <li>store all its results on the passed {@code blackboard}.
+	 * <li>assure that if {@link #canContribute(ReadOnlyBlackboardView)} is called with
```
The list is formatted horrible.

acBerger commented:
@jGleitz Please resolve conflicts

acBerger commented:
Currently checkstyle is turned off here, isn't it? This should be changed before merging.

acBerger commented on src/main/java/de/uka/sdq/beagle/core/BeagleController.java:
```diff
@@ -0,0 +1,23 @@
+	 * data exchange in any action invoked by this controller must happen exclusively via
+	 * this blackboard instance.
+	 */
+	// private final Blackboard blackboard = new Blackboard();
```
why is that?

acBerger commented:
> Currently checkstyle is turned off here, isn't it? This should be changed before merging.

sorry, I was confused. It is correct as it is.

acBerger commented:
I found some things which would not work with Checkstyle activated but I think this should be corrected, when implementing.

lgtm.

jGleitz commented:
* fixed package path
* improved `<ul>` formatting (I see no other way than using these newlines because of the Eclipse Formatter)
* Added my comment as documentation to `BlackboardStorer`.
